[CORE-13837] ct: Add level zero metrics

Lazin opened the pull request:
Add pipeline metrics.
The metrics include things like total number of requests in/out/errored, total number of bytes flowing in/out, request latency histogram, request size histogram, etc.

The probe is generic and used for both read pipeline and write pipeline.
The arch. of the pipeline allows to collect all metrics at the point of request ingestion (e.g. write_and_debounce method). There is no need to sprinkle probe calls around the code which makes code a bit simpler. Some pipeline components need their own micro probes. 

All L0 probes are located in one place (`src/v/cloud_topics/level_zero/common/level_zero_probe`) for convenience. 

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->
* none

Lazin committed:
ct: Add pipeline probe

The probe can be used for both read and write pipelines.

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

Lazin committed:
ct: Use pipeline probe in write pipeline

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

Lazin committed:
ct: Use pipeline probe in the read pipeline

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--2c35c7c2-ddf3-4789-9eba-7c90995c7d31-->
<!--TEST RESULTS-->
<!--db7a108b-28f6-4d7d-a6da-b4cba54351b1-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/73615">build#73615</a></summary>


| test_class                 | test_method                              | test_arguments                                                                                                                          | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  | test_history                                                                                                                                                                      |
|:---------------------------|:-----------------------------------------|:----------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| RandomNodeOperationsTest   | test_node_operations                     | {"cloud_storage_type": 1, "compaction_mode": "sliding_window", "enable_failures": false, "mixed_versions": true, "with_iceberg": false} | integration | https://buildkite.com/redpanda/redpanda/builds/73615#0199b495-ca1e-4e7a-9320-1b092b425712 | FLAKY         | 19/21    | upstream reliability is '98.39357429718876'. current run reliability is '90.47619047619048'. drift is 7.91738 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RandomNodeOperationsTest&test_method=test_node_operations                       |
| TimeQueryTest              | test_timequery_with_local_gc             | null                                                                                                                                    | integration | https://buildkite.com/redpanda/redpanda/builds/73615#0199b495-ca11-404b-860c-7695acfcaaa1 | FLAKY         | 17/21    | upstream reliability is '92.18934911242603'. current run reliability is '80.95238095238095'. drift is 11.23697 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=TimeQueryTest&test_method=test_timequery_with_local_gc                          |
| TxAtomicProduceConsumeTest | test_basic_tx_consumer_transform_produce | {"with_failures": true}                                                                                                                 | integration | https://buildkite.com/redpanda/redpanda/builds/73615#0199b495-ca13-41e1-85b3-9cc5233a15b1 | FLAKY         | 20/21    | upstream reliability is '99.29245283018868'. current run reliability is '95.23809523809523'. drift is 4.05436 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=TxAtomicProduceConsumeTest&test_method=test_basic_tx_consumer_transform_produce |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/73640">build#73640</a></summary>


| test_class                                         | test_method                                        | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason   | test_history   |
|:---------------------------------------------------|:---------------------------------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:---------|:---------------|
| src/v/cloud_storage/tests/segment_meta_cstore_test | src/v/cloud_storage/tests/segment_meta_cstore_test |                  | unit        | https://buildkite.com/redpanda/redpanda/builds/73640#0199ba51-4ad7-4a80-91fe-32fd0135fa20 | FAIL          | 0/1      |          |                |
</details>
<!--/body-->

Lazin committed:
ct: Add throttler probe

The probe tracks number of throttled requests (counter and gauge) and
total memory being throttled at the moment.

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

Lazin committed:
ct: Integrate new probe into throttler

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

Lazin committed:
ct: Add write_request_scheduler_probe

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

Lazin committed:
ct: Use probe in the write_request_scheduler

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

Lazin committed:
ct: Add batcher probe

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

Lazin committed:
ct: Use probe in the batcher

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

rockwotj reviewed (approved):
LGTM

dotnwat reviewed (commented):
👍
